File: 3. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int res = 0;
        vector<int> dict(256, -1);
        int i = 0, start = 0;
        for (; i < s.size(); i++) {
            start = max(start, dict[s[i]] + 1);
            dict[s[i]] = i;
            res = max(res, i - start + 1);
        }
        
        return res;
    }
};

File: go/src/leetcode/3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.go
package main

import "fmt"

/**
https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/
 */
func lengthOfLongestSubstring(s string) int {
	var l, last int
	mapper := make(map[byte]int)
	for i := range s {
		c := s[i]
		if j, ok := mapper[c]; ok {
			last = max(j + 1, last)
		}
		l = max(l, i - last + 1)
		mapper[c] = i
	}
	return l
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main()  {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
	fmt.Println(lengthOfLongestSubstring("bbbbb"))
	fmt.Println(lengthOfLongestSubstring("pwwkew"))
	fmt.Println(lengthOfLongestSubstring(""))
	fmt.Println(lengthOfLongestSubstring(" "))
	fmt.Println(lengthOfLongestSubstring("abc"))
	fmt.Println(lengthOfLongestSubstring("abba"))
}
